proposals/lip-introduce-reward-sharing

```diff
@@ -0,0 +1,549 @@
+
+
+```python
+def calculateVoteRewards(vote: Vote,tokenID: TokenID):
+    i = index of item in delegateStore(vote.delegateAddress).sharingCoefficients with item.tokenID == tokenID
+
+    # transform variables to Q96    
+    delegateSharingCoefficient = bytesToQ96(delegateStore(vote.delegateAddress).sharingCoefficients[i].coefficient)
+    amount = Q96(vote.amount)
+    voteSharingCoefficient = bytesToQ96(vote.voteSharingCoefficients[i].coefficient)
+
+    # calculate reward
+    reward = mul_96(amount, sub_96(delegateSharingCoefficient, voteSharingCoefficient))
+
+    return Q_96_ToInt(reward)
+```
+
+Note that here `i` specifies the index of entries for token with id `tokenID` in the sharing coefficients array in both for the delegate and the voter substore. The fact that both arrays have the same `tokenID` in the same position is guaranteed from the fact that both arrays are sorted in lexicographic order. For the delegate substore array this is done in the [`updateSharedRewards`](#updatesharedrewards) function and for the voter substore in the [assignVoteRewards](#assignvoterewards) function.
+
+#### assignVoteRewards
+
+This function assigns the rewards to the specified voter for a specific vote, the `i`th vote stored in the `sentVotes` array of `voterStore(address)`.
+
+```python
+def assignVoteRewards(voterAddress: Address, i: uint32) -> None:
+    vote = voterStore(voterAddress).sentVotes[i]
+    # self-votes are excluded from reward sharing
+    if vote.delegateAddress == voterAddress:
+        return
+
+    # assign rewards for each token separately
+    for elem in delegateStore(vote.delegateAddress).sharingCoefficients:
+        if there does not exist an item in vote.voteSharingCoefficients with item.tokenID == elem.tokenID:
+            vote.voteSharingCoefficients.append({"tokenID": item.tokenID, "coefficient": q96ToBytes(Q96(0))})  
+            keep the vote.voteSharingCoefficients array ordered in lexicographic order of tokenID 
+            # this makes sure that the order is the same as in delegate substore
+
+        tokenID = elem.tokenID
+        reward = calculateVoteRewards(vote, tokenID)
+
+        if reward > 0:
+            # unlock and send tokens to voter
+            Token.unlock(vote.delegateAddress,
+                     	   MODULE_NAME_DPOS,
+                    	   tokenID,
+                    	   reward
+                    	)    
+            Token.transfer(vote.delegateAddress, 
+                          voterAddress,
+                          tokenID,
+                          reward)
+
+            emitEvent(
+                module = MODULE_NAME_DPOS,
+                name = EVENT_NAME_REWARDS_ASSIGNED,
+                data={
+                    "voterAddress": voterAddress,
+                    "delegateAddress": vote.delegateAddress,
+                    "tokenID": tokenID,
+                    "amount": reward
+                },
+                topics = [voterAddress]
+            )
```

Comment:
As far as I see we now move the update of the vote sharing coefficient outside this function, so it needs to happen outside of this function after/before this function is called. Is there any specific reason for this specification change?

Reply:
It needs to happen after the function is called actually.

This change would not make sense from my point of view; since this has to be done *every* time after transferring the rewards to the voter, it is less error prone to include it inside the function. However the SDK team preferred to keep separate the store update from this function and just perform the unlock and transfer here. Then, we should update the state store every time after calling this function. The change happened to align the LIP with the actual implementation.

Reply:
Okay, let us keep it as is even though I also agree that the previous specs were cleaner.

In general, I think it is fine if implementation and specifications differ in how functions are structures as long as the protocol is the same in the end. Therefore, we should not feel obliged to change it.

Reply:
ok, it is easy (and perhaps better) for me to revert it back.

Reply:
To avoid another review, I would keep it as is now.

Reply:
If you prefer to make the change, feel free to do so. I will review again :-)